rosbag2_cpp/src/rosbag2_cpp/clocks/time_controller_clock.cpp

```diff
@@ -0,0 +1,107 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <mutex>
+#include <thread>
+
+#include "rcpputils/thread_safety_annotations.hpp"
+#include "rosbag2_cpp/clocks/time_controller_clock.hpp"
+#include "rosbag2_cpp/types.hpp"
+
+namespace rosbag2_cpp
+{
+
+class TimeControllerClockImpl
+{
+public:
+  /**
+   * Stores an exact time match between a system steady clock and the playback ROS clock.
+   * This snapshot is taken whenever a factor changes such that a new reference is needed,
+   * such as pause, resume, rate change, or jump
+   */
+  struct TimeReference
+  {
+    PlayerClock::ROSTimePoint ros;
+    PlayerClock::SteadyTimePoint steady;
+  };
+
+  TimeControllerClockImpl() = default;
+  virtual ~TimeControllerClockImpl() = default;
+
+  template<typename T>
+  rcutils_duration_value_t duration_nanos(const T & duration)
+  {
+    return std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count();
+  }
+
+  PlayerClock::ROSTimePoint steady_to_ros(PlayerClock::SteadyTimePoint steady_time)
+  {
+    return reference.ros + (rate * duration_nanos(steady_time - reference.steady));
+  }
+
+  PlayerClock::SteadyTimePoint ros_to_steady(PlayerClock::ROSTimePoint ros_time)
+  {
+    const rcutils_duration_value_t diff_nanos = (ros_time - reference.ros) / rate;
+    return reference.steady + std::chrono::nanoseconds(diff_nanos);
+  }
+
+  double rate = 1.0;
+  PlayerClock::NowFunction now_fn;
+  std::mutex mutex;
+  TimeReference reference RCPPUTILS_TSA_GUARDED_BY(mutex);
+};
+
+TimeControllerClock::TimeControllerClock(
+  ROSTimePoint starting_time,
+  double rate,
+  NowFunction now_fn)
+: impl_(std::make_unique<TimeControllerClockImpl>())
+{
+  std::lock_guard<std::mutex> lock(impl_->mutex);
+  impl_->now_fn = now_fn;
+  impl_->reference.ros = starting_time;
+  impl_->reference.steady = impl_->now_fn();
+  impl_->rate = rate;
+}
+
+TimeControllerClock::~TimeControllerClock()
+{}
+
+PlayerClock::ROSTimePoint TimeControllerClock::now() const
+{
+  std::lock_guard<std::mutex> lock(impl_->mutex);
+  return impl_->steady_to_ros(impl_->now_fn());
+}
+
+bool TimeControllerClock::sleep_until(ROSTimePoint until)
+{
+  SteadyTimePoint steady_until;
+  {
+    std::lock_guard<std::mutex> lock(impl_->mutex);
+    steady_until = impl_->ros_to_steady(until);
+  }
+  // TODO(emersonknapp) - when we have methods that can change timeflow during a sleep,
+  // it will probably be better to use a condition_variable::wait_until
+  std::this_thread::sleep_until(steady_until);
+  return now() >= until;
```

Comment:
what is the consequence for the Player if that sleep_until fails?

Reply:
they sleep again - you may notice that Player runs `while (rclcpp::ok() && !sleep_until(message->timestamp)) {}`

Reply:
I may be off here, but consider this:

* dequeue message at time `t0+10`
* `while !sleep_until...`
* <external change of rate or playback, time now yields `t0+5`.

We should actually load & dequeue another message from the bag. Otherwise we'd sleep an additional 5 seconds skipping over existing messages in the bag until we hit `t0+10` again.

Reply:
That example doesn't parse as a possibility for me the way I am reading it, let me put it back to you the way I understood it:
* dequeue message, where `message->time_stamp == t0+10`
* `while !sleep_until...`
* External source calls `set_rate` - so we should be sleeping for less time than originally anticipated

What happens now is that when the external source calls `set_rate()`, the `condition_variable.notify_all` is called, so the player awakens immediately (sleep == `condition_variable.wait_until`), and will sleep again until new appropriate time

Or, are you talking about `jump` - which can move us to an arbitrary time, and therefore moves us back in time? This absolutely needs a "queue-busting" effect, as called out in the design documents via jump callbacks, which will also need to clear the most-recently-dequeued variable here. I would say that needs to be carefully considered during the implementation of `jump`.

Reply:
Note: `PlayerClock::now()` is monotonically increasing in all cases unless `jump` is called, therefore the "next messaage" always stays correct (unless jump is called) - and this call will have to trigger extra logic in the `Player` - it cannot operate in the `PlayerClock` alone. 

I believe `jump` will therefore also need new `Reader` APIs - or use the approach "start over from the beginning and dequeue messages up to the appropriate time", which is probably not the desired approach.

Reply:
my bad, I wasn't verbalizing my thought correctly. I was indeed referring to `jump` rather than `rate`. 
I'll take it from your comments that we currently don't consider a time jump and therefore don't cope with `now()` being potentially a time stamp in the past.

Reply:
> we currently don't consider a time jump and therefore don't cope with now() being potentially a time stamp in the past.

Yeah, that's correct. Handling that will definitely need some updates here, but I'm not sure precisely what they will look like yet